tools/donate_cpu_lib.py

```diff
@@ -135,20 +135,22 @@ def get_packages_count(server_address):
 
 
 def get_package(server_address, package_index=None):
-    print('Connecting to server to get assigned work..')
     package = b''
-    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-    try:
-        sock.connect(server_address)
-        if package_index is None:
-            sock.send(b'get\n')
-        else:
-            request = 'getPackageIdx:' + str(package_index) + '\n'
-            sock.send(request.encode())
-        package = sock.recv(256)
-    except socket.error:
-        pass
-    sock.close()
+    while not package:
+        print('Connecting to server to get assigned work..')
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        try:
+            sock.connect(server_address)
+            if package_index is None:
+                sock.send(b'get\n')
+            else:
+                request = 'getPackageIdx:' + str(package_index) + '\n'
+                sock.send(request.encode())
+            package = sock.recv(256)
+        except socket.error:
+            print("network or server might be temporarily down.. will try again in 30 seconds..")
+            time.sleep(30)
+        sock.close()
```

Comment:
not sure how to handle this best. It seems unfortunate to have the `sock.close()` after the sleep.

what do you think about a `with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:` inside the try?

I'm not a python+network expert so well I don't know what is best.

Reply:
> what do you think about a `with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:` inside the try?

Looks much better, thanks.

I believe there is room for using `with socket.socket(...) as sock:`, for example below, were I don't believe we close the socket on failure (otoh, the script exits after that so I guess it's cleaned up right away anyway)? Changing that would be a separate PR though.
https://github.com/danmar/cppcheck/blob/206eb0f5271716dc863194a3ab1223749a99ebbb/tools/donate_cpu_lib.py#L111-L119